### PR TITLE
Add getcmdprompt()

### PR DESCRIPTION
**Pull request by Shougo**
I have added `getcmdprompt()` to get command line prompt.

### CODE REVIEW

**Comment by k-takata** [comment]
`:help builtin-function-list` should be also updated.

File: src/ex_getln.c [hunk 1]
```diff
@@ -4230,6 +4230,25 @@ f_getcmdline(typval_T *argvars UNUSED, typval_T *rettv)
...
+    cmdline_info_T *p = get_ccline_ptr();
+    rettv->v_type = VAR_STRING;
+    if (p != NULL)
+    {
+	char_u *msg = msg_get();
+	int msg_len = (int)STRLEN(msg);
+	int new_len = msg_len - p->cmdlen;
+	rettv->vval.v_string =
+	    (new_len > 0) ? vim_strnsave(msg, new_len) : NULL;
+    } else
```
Comment (k-takata) [inline src/ex_getln.c]:
```suggestion
    }
    else
```

File: src/testdir/test_cmdline.vim [hunk 2]
```diff
@@ -1641,6 +1641,11 @@ func Test_getcmdtype()
   cunmap <F6>
 
   call assert_equal('', getcmdline())
+
+  call assert_equal('', getcmdmsg())
+  autocmd CmdlineChanged * call assert_equal('Answer?', getcmdmsg())
+  call feedkeys(":call input('Answer?')\<CR>", "t")
```
Comment (Shougo) [inline src/testdir/test_cmdline.vim]: The test seems not executed.
How to execute `input()` in tests?

Comment (zeertzjq) [reply src/testdir/test_cmdline.vim]: Change `"t"` to `"tx"`

Comment (Shougo) [reply src/testdir/test_cmdline.vim]: Fixed.

**Comment by Shougo** [comment]
I don't know how to fix the error...

```
==15384==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x502000057d36 at pc 0x55e963dd5b03 bp 0x7ffe50d38dd0 sp 0x7ffe50d38598
READ of size 7 at 0x502000057d36 thread T0
    #0 0x55e963dd5b02 in ___interceptor_strlen ??:?
    #1 0x55e964d3b5df in msg_puts_attr_len /home/runner/work/vim/vim/src/message.c:2272:9
    #2 0x55e964d3a5c7 in msg_outtrans_len_attr /home/runner/work/vim/vim/src/message.c:1723:2
    #3 0x55e9641c20bd in draw_cmdline /home/runner/work/vim/vim/src/ex_getln.c:3474:2
    #4 0x55e9641c4059 in put_on_cmdline /home/runner/work/vim/vim/src/ex_getln.c:3632:6
    #5 0x55e9641b9168 in getcmdline_int /home/runner/work/vim/vim/src/ex_getln.c
    #6 0x55e964146880 in do_cmdline /home/runner/work/vim/vim/src/ex_docmd.c:911:46
    #7 0x55e96444ac23 in nv_colon /home/runner/work/vim/vim/src/normal.c:3166:15
    #8 0x55e964428aa0 in normal_cmd /home/runner/work/vim/vim/src/normal.c:949:5
    #9 0x55e96417ef0c in exec_normal /home/runner/work/vim/vim/src/ex_docmd.c
    #10 0x55e9640ad321 in f_feedkeys /home/runner/work/vim/vim/src/evalfunc.c:4749:6
    #11 0x55e96409e3c5 in call_internal_func /home/runner/work/vim/vim/src/evalfunc.c:3275:5
    #12 0x55e964932452 in call_func /home/runner/work/vim/vim/src/userfunc.c:4016:14
    #13 0x55e964930d54 in get_func_tv /home/runner/work/vim/vim/src/userfunc.c:2038:8
    #14 0x55e96495c03b in ex_call_inner /home/runner/work/vim/vim/src/userfunc.c:6227:6
    #15 0x55e96495c03b in ex_call /home/runner/work/vim/vim/src/userfunc.c:6575:11
    #16 0x55e96414fa7b in do_one_cmd /home/runner/work/vim/vim/src/ex_docmd.c:2616:2
    #17 0x55e964147577 in do_cmdline /home/runner/work/vim/vim/src/ex_docmd.c:1028:17
    #18 0x55e96493b09f in call_user_func /home/runner/work/vim/vim/src/userfunc.c:3201:2
    #19 0x55e964938590 in call_user_func_check /home/runner/work/vim/vim/src/userfunc.c:3374:10
    #20 0x55e96493289b in call_func /home/runner/work/vim/vim/src/userfunc.c:3998:15
    #21 0x55e964930d54 in get_func_tv /home/runner/work/vim/vim/src/userfunc.c:2038:8
    #22 0x55e96495c03b in ex_call_inner /home/runner/work/vim/vim/src/userfunc.c:6227:6
    #23 0x55e96495c03b in ex_call /home/runner/work/vim/vim/src/userfunc.c:6575:11
    #24 0x55e96414fa7b in do_one_cmd /home/runner/work/vim/vim/src/ex_docmd.c:2616:2
    #25 0x55e964147577 in do_cmdline /home/runner/work/vim/vim/src/ex_docmd.c:1028:17
    #26 0x55e96408a68a in ex_execute /home/runner/work/vim/vim/src/eval.c:7524:6
    #27 0x55e96414fa7b in do_one_cmd /home/runner/work/vim/vim/src/ex_docmd.c:2616:2
    #28 0x55e964147577 in do_cmdline /home/runner/work/vim/vim/src/ex_docmd.c:1028:17
    #29 0x55e96493b09f in call_user_func /home/runner/work/vim/vim/src/userfunc.c:3201:2
    #30 0x55e964938590 in call_user_func_check /home/runner/work/vim/vim/src/userfunc.c:3374:10
    #31 0x55e96493289b in call_func /home/runner/work/vim/vim/src/userfunc.c:3998:15
    #32 0x55e964930d54 in get_func_tv /home/runner/work/vim/vim/src/userfunc.c:2038:8
    #33 0x55e96495c03b in ex_call_inner /home/runner/work/vim/vim/src/userfunc.c:6227:6
    #34 0x55e96495c03b in ex_call /home/runner/work/vim/vim/src/userfunc.c:6575:11
    #35 0x55e96414fa7b in do_one_cmd /home/runner/work/vim/vim/src/ex_docmd.c:2616:2
    #36 0x55e964147577 in do_cmdline /home/runner/work/vim/vim/src/ex_docmd.c:1028:17
    #37 0x55e9646a9153 in do_source_ext /home/runner/work/vim/vim/src/scriptfile.c:1806:5
    #38 0x55e9646a65be in do_source /home/runner/work/vim/vim/src/scriptfile.c:1952:12
    #39 0x55e9646a65be in cmd_source /home/runner/work/vim/vim/src/scriptfile.c:1298:14
    #40 0x55e96414fa7b in do_one_cmd /home/runner/work/vim/vim/src/ex_docmd.c:2616:2
    #41 0x55e964147577 in do_cmdline /home/runner/work/vim/vim/src/ex_docmd.c:1028:17
    #42 0x55e964d1d6b4 in exe_commands /home/runner/work/vim/vim/src/main.c:3173:2
    #43 0x55e964d1d6b4 in vim_main2 /home/runner/work/vim/vim/src/main.c:790:2
    #44 0x55e964d16219 in main /home/runner/work/vim/vim/src/main.c:441:12
    #45 0x7f6c9b029d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #46 0x7f6c9b029e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #1 0x55e963dbdf94 in _start ??:?
```

**Comment by Shougo** [comment]
Fixed.

**Comment by chrisbra** [comment]
Thanks, but I am not sure I understand this yet. What is meant with:
> Return the current command-line message. 

What message? or what command-prompt are you talking about? Can you please give an example of where you need this?

**Comment by Shougo** [comment]
I want to get the command line prompt like this.

```vim
call input('foo')
```

The current input can be get by `getcmdline()`, but the prompt `foo` is cannot be get.

**Comment by zeertzjq** [comment]
The approach in this PR can only get the last line if the prompt is multi-line (e.g. `input("abc\ndef\nghi")`). Is that intended? If not, maybe it's better to save the message in `get_user_input()`.

**Comment by Shougo** [comment]
> The approach in this PR can only get the last line if the prompt is multi-line (e.g. input("abc\ndef\nghi")). Is that intended? If not, maybe it's better to save the message in get_user_input().

I have changed the implementation.
But how to test the function?

File: src/evalfunc.c [hunk 3]
```diff
@@ -3776,6 +3778,7 @@ f_confirm(typval_T *argvars UNUSED, typval_T *rettv UNUSED)
 	return;
 
     message = tv_get_string_chk(&argvars[0]);
+    set_prompt(message);
```
Comment (zeertzjq) [inline src/evalfunc.c]: This should come after the NULL check, otherwise it will crash on type error.

File: src/ex_getln.c [hunk 4]
```diff
@@ -4853,6 +4886,7 @@ get_user_input(
 	return;
 
     prompt = tv_get_string_chk(&argvars[0]);
+    set_prompt(prompt);
```
Comment (zeertzjq) [inline src/ex_getln.c]: ditto

File: src/evalfunc.c (hunk 3 above)
Comment (Shougo) [reply src/evalfunc.c]: Fixed.

File: src/ex_getln.c (hunk 4 above)
Comment (Shougo) [reply src/ex_getln.c]: Fixed.

File: src/testdir/test_cmdline.vim [hunk 5]
```diff
@@ -1641,6 +1641,25 @@ func Test_getcmdtype()
   cunmap <F6>
 
   call assert_equal('', getcmdline())
+
+  call assert_equal('', getcmdprompt())
+  augroup test_CmdlineEnter
+    autocmd!
+    autocmd CmdlineEnter * call assert_equal('Answer?', getcmdprompt())
```
Comment (chrisbra) [inline src/testdir/test_cmdline.vim]:
```suggestion
    autocmd CmdlineEnter * let g:cmdprompt=getcmdprompt()
```

File: src/testdir/test_cmdline.vim [hunk 6]
```diff
@@ -1641,6 +1641,25 @@ func Test_getcmdtype()
...
+  augroup END
+  call feedkeys(":call input('Answer?')\<CR>a\<CR>\<ESC>", "xt")
+  call assert_equal('', getcmdprompt())
```
Comment (chrisbra) [inline src/testdir/test_cmdline.vim:1652]:
```suggestion
  call assert_equal('Answer?', g:cmdprompt)
  call assert_equal('', getcmdprompt())
```

File: src/testdir/test_cmdline.vim [hunk 7]
```diff
@@ -1641,6 +1641,25 @@ func Test_getcmdtype()
...
+
+  augroup test_CmdlineEnter
+    autocmd!
+    autocmd CmdlineEnter * call assert_equal('confirm?', getcmdprompt())
+  augroup END
+  call feedkeys(":call confirm('confirm?')\<CR>o\<ESC>", "xt")
+
```
Comment (chrisbra) [inline src/testdir/test_cmdline.vim]:
```suggestion
   call assert_equal('Answer?', g:cmdprompt)
   call assert_equal('', getcmdprompt())
```

File: src/testdir/test_cmdline.vim [hunk 8]
```diff
@@ -1641,6 +1641,25 @@ func Test_getcmdtype()
...
+    autocmd!
+    autocmd CmdlineEnter * call assert_equal('Answer?', getcmdprompt())
+  augroup END
+  call feedkeys(":call input('Answer?')\<CR>a\<CR>\<ESC>", "xt")
+  call assert_equal('', getcmdprompt())
+
+  augroup test_CmdlineEnter
+    autocmd!
+    autocmd CmdlineEnter * call assert_equal('confirm?', getcmdprompt())
+  augroup END
```
Comment (chrisbra) [inline src/testdir/test_cmdline.vim]: You can get rid of that block now

File: src/ex_getln.c [hunk 9]
```diff
@@ -4230,6 +4234,24 @@ f_getcmdline(typval_T *argvars UNUSED, typval_T *rettv)
...
+    return current_prompt;
+}
+
+/*
+ * Set current command line prompt.
+ */
+    void
+set_prompt(char_u* str)
+{
+    vim_strncpy(current_prompt, str, sizeof(current_prompt) - 1);
```
Comment (chrisbra) [inline src/ex_getln.c:4252]: I think you need to free current_prompt when it is non-null before setting it.

File: src/testdir/test_cmdline.vim (hunk 8 above)
Comment (Shougo) [reply src/testdir/test_cmdline.vim]: Thanks.

File: src/ex_getln.c (hunk 9 above)
Comment (Shougo) [reply src/ex_getln.c:4252]:
```c
static char_u	current_prompt[CMDBUFFSIZE + 1] = "";
```

`current_prompt` is static array.  Why free is needed?

**Comment by Shougo** [comment]
I need to input when execute `test_cmdline`.
Hm...

File: src/ex_getln.c (hunk 9 above)
Comment (chrisbra) [reply src/ex_getln.c:4252]: ah, I thought this allocates memory, but it uses strcpy, so I guess this should be fine.

**Pull request closed without merging**